### PR TITLE
[lang] [refactor] Add a ExtArray class for external arrays

### PR DESCRIPTION
**Pull request by strongoier**
Related issue = #

Previously, things around external arrays are implemented under `Expr` class, which is not intuitive because `Expr` class should represent a scalar supporting `TaichiOperations`. This PR mostly involves slight code migration and internal API changes. After this PR, `Expr` class will become (hopefully) clean.

### CODE REVIEW

**Review by ailzhang — COMMENTED** [review]
Thanks for the PR! Overall decoupling ExtArr with Expr makes sense to me, might come back later today with more questions/comments :D
Update: IIUC we don't have torch installed in our CI environment (by grepping torch in the .github folder). If this is true it's better that we enable the tests first before merging this PR. ;)

File: python/taichi/lang/ext_array.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+from taichi.core.util import ti_core as _ti_core
+from taichi.lang.expr import Expr
+from taichi.lang.util import taichi_scope
+
+
+class ExtArray:
+    """Class for external arrays. Constructed by a C++ Expr wrapping a C++ ExternalTensorExpression.
+
+    Args:
+        ptr: The C++ Expr mentioned above.
```
[the remark this answers is not in the document]
Comment (strongoier) [reply python/taichi/lang/ext_array.py:10]: Yes I should use `taichi_core.Expr` in the docstring. But is there a good way to name things without a binding like a C++ `ExternalTensorExpression`?

File: python/taichi/lang/impl.py [hunk 2]
```diff
@@ -72,15 +73,16 @@ def expr_init_func(
 
 
 def begin_frontend_struct_for(group, loop_range):
-    if not isinstance(loop_range, Expr) or not loop_range.is_global():
-        raise TypeError('Can only iterate through global variables/fields')
+    if not isinstance(loop_range, (ExtArray, Field, SNode, _Root)):
+        raise TypeError(
+            'Can only iterate through Taichi fields or external arrays')
```
[the remark this answers is not in the document]
Comment (strongoier) [reply python/taichi/lang/impl.py:78]: Oh you've pointed out a nice problem. It's my fault that I just kept the check statement same as before assuming it was implemented correctly. However, struct fors on external arrays are not supported in Taichi yet.

File: python/taichi/lang/ext_array.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ailzhang) [reply python/taichi/lang/ext_array.py:10]: Oh IMHO it's okay to keep the binding name `taichi_core` as long as it matches your C++ extension name. You might have multiple C++ extensions in the future so having the extension name there is helpful too ;)

File: python/taichi/lang/impl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ailzhang) [reply python/taichi/lang/impl.py:78]: Oh in that case shall we update the logic here? At least the error message indicates iterating through external array is supported.

**Review by k-ye — COMMENTED** [review]
LGTM (I'll leave @ailzhang to decide whether this is intuitive enough :D)

**Comment by strongoier** [comment]
> LGTM (I'll leave @ailzhang to decide whether this is intuitive enough :D)

Just as @ailzhang mentioned, are we going to install torch in our CI environment?

**Comment by k-ye** [comment]
> are we going to install torch in our CI environment?

Sure

File: python/taichi/lang/impl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (strongoier) [reply python/taichi/lang/impl.py]: done

File: python/taichi/lang/ext_array.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (strongoier) [reply python/taichi/lang/ext_array.py]: done

File: python/taichi/lang/ext_array.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        """
+        dim = _ti_core.get_external_tensor_dim(self.ptr)
+        ret = [
+            Expr(_ti_core.get_external_tensor_shape_along_axis(self.ptr, i))
+            for i in range(dim)
+        ]
+        return ret
+
+    @taichi_scope
+    def loop_range(self):
```
Comment (ailzhang) [inline python/taichi/lang/ext_array.py:32]: I wonder if we still need this if we don't support struct for :P

**Review by ailzhang — COMMENTED** [review]
This PR looks good to me! two minor nits before we can merge this:
1. It's a bit unclear to me why we create a separate class in python while keeping the C++ class inheritance (ExternalTensorExpression and  Expr). It'd be nice if you can add some context about the motivation of this PR. 
2. Enabling torch tests in CI so that we can test this PR before merging :D

File: python/taichi/lang/ext_array.py (hunk 3 above)
Comment (strongoier) [reply python/taichi/lang/ext_array.py:32]: we'll support struct for in the future :-)

Comment (ailzhang) [reply python/taichi/lang/ext_array.py:32]: lol sure - maybe add a comment here saying that it's actually not supported? With this method it looks pretty valid :P

Comment (strongoier) [reply python/taichi/lang/ext_array.py:32]: done

**Review by ailzhang — APPROVED** [review]
Looks great, thanks @strongoier ! Please feel free to merge once CI is green.

**Pull request merged**